pallets/auctions/src/lib.rs

```diff
@@ -0,0 +1,726 @@
+                    .ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+                // ensure the auction period has not ended
+                ensure!(auction.end_block > current_block, Error::<T>::AuctionEnded);
+
+                // ensure the bid is larger than the current highest bid
+                if let Some(current_highest_bid) = auction.bidders.get_highest_bid() {
+                    ensure!(amount > current_highest_bid.1, Error::<T>::InvalidBidAmount);
+                }
+
+                let user_bid = auction
+                    .bidders
+                    .remove_bid(who.clone())
+                    .ok_or(Error::<T>::BidDoesNotExist)?;
+
+                let amount_to_transfer = amount
+                    .checked_sub(&user_bid.1)
+                    .ok_or(Error::<T>::UnexpectedError)?;
+
+                T::Currency::transfer(&who, &Self::account_id(), amount_to_transfer, KeepAlive)?;
+
+                auction.bidders.insert_new_bid(who.clone(), amount);
+
+                // emit bid updated event
+                Self::deposit_event(Event::BidUpdated {
+                    nft_id,
+                    marketplace_id: auction.marketplace_id,
+                    creator: who,
+                });
+                Ok(())
+            })?;
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        #[transactional]
+        pub fn buy_it_now(
+            origin: OriginFor<T>,
+            nft_id: NFTId,
+            amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let current_block = frame_system::Pallet::<T>::block_number();
+
+            // transfer funds from caller to pallet
+            T::Currency::transfer(&who, &Self::account_id(), amount, KeepAlive)?;
+
+            // Mark the auction as completed so other bid users can claim amount
+            Auctions::<T>::try_mutate(nft_id, |maybe_auction| -> DispatchResult {
+                // should not panic when unwrap since already checked above
+                let mut auction = maybe_auction
+                    .as_mut()
+                    .ok_or(Error::<T>::AuctionDoesNotExist)?;
+                // ensure the auction has a buy it now price
+                ensure!(
+                    auction.buy_it_price.is_some(),
+                    Error::<T>::AuctionDoesNotSupportBuyItNow
+                );
+
+                // ensure the auction period has commenced
+                ensure!(
+                    auction.start_block < current_block,
+                    Error::<T>::AuctionNotStarted
+                );
+
+                // ensure the auction period has not ended
+                ensure!(auction.end_block > current_block, Error::<T>::AuctionEnded);
+                auction.state = AuctionState::Completed;
+
+                // transfer NFT to caller
+                T::NFTHandler::set_owner(nft_id, &who)?;
```

Comment:
First we do the checking and then we do our work. This needs to be at the end right before we emit an event.